Add ERC: API for Hierarchical Accounts (#932)
ethereum/ERCs: ERCS/erc-7895.md

---
eip: 7895
title: API for Hierarchical Accounts
description: Adds JSON-RPC method for requesting that a universal wallet create or track another account that it owns
author: Wilson Cusack (@wilsoncusack), Jake Feldman (@jakefeldman), Montana Wong (@montycheese), Felix Zhang (@fan-zhang-sv), Jake Moxey (@jxom)
discussions-to: https://ethereum-magicians.org/t/wallet-addsubaccount/23013
status: Draft
type: Standards Track
category: ERC
created: 2025-02-18
---

## Abstract

> This ERC introduces a new wallet RPC, `wallet_addSubAccount`, which allows an app to request a wallet track a smart account that the wallet owns. It also allows apps to request the wallet to provision a new account, owned by the universal wallet with a signer provided by the caller. 

Reviewer: I still think you need more details about what "track" means in this context. I assume that the idea is that the dapp wants a unique address, and "track" means both "display in the wallet UI" and "allow the user to send transactions from it"?

## Motivation

> Embedded app accounts (onchain accounts specific to a single app) have led to a proliferation of user addresses, which can be difficult for users to keep track of. Many embedded app account users also have a universal wallet, which can be used across apps. With hierarchical ownership–where one smart account can own another–if the embedded app account is a smart account, it could be owned by the user’s universal wallet. This would allow users to be able to control an app account via their universal wallet. However, though hierarchical ownership is already possible today, there is no way for apps to tell universal wallets about embedded app accounts a user may have. The proposed RPC provides a path for this. 

Reviewer: Should define what a "universal wallet" is. Something like MetaMask that you use cross-dapp?

}

// Requester is agnostic to account type and only wants to ensure its signer is an owner
type CreateAccount = {
  type: "create";
  keys: { 
    publicKey: "0x...";
    type: "address" | "p256" | "webcrypto-p256" |  "webauthn-p256";
  }[];
}

// Undeployed account, app creates the account
type UndeployedAccount = {
  type: "undeployed";
  address: `0x${string}`;
  keys: never;
  chainId?: '0x{string}'; // in Hex
  // Required: factory address to create the account
  factory: `0x${string}`;
  // Required: factory calldata for the account
  factoryData: `0x${string}`;
}

type Request = {
  method: "wallet_addSubAccount";
  params: [{
    // JSON-RPC method version
    version: string;
    // JSON-RPC method account 
    account: CreateAccount | DeployedAccount | UndeployedAccount;
  }],
}
```

##### Response

Factory data and factory address are OPTIONAL and SHOULD be returned when available. Deployed accounts MAY not have these fields.

```typescript
type Response = {
  // Address of the account.
  address: `0x${string}`;
  // Optional: factory address
  factory?: `0x${string}`;
  // Optional: factory calldata
  factoryData?: `0x${string}`;
}
```

##### `CreateAccount`

> Allows the wallet to create a Sub Account with a list of known signers. This enables applications to provide a signer for the new account. A wallet SHOULD make the universal account an owner of the account, creating a hierarchical relationship between the newly created account and the universal account.
> 
> ```typescript
> type Parameters = {
>   address: never;
>   // Required: keys of the account to be created
>   keys: { 
>     publicKey: "0x...";
>     type: "address" | "p256" | "webcrypto-p256" |  "webauthn-p256";
>   }[];
>   factory: never;
>   factoryData: never;
> }
> ```

Reviewer 2: Should we push for the Wallet's Provider to create & manage keys by default here? 

```suggestion
Creates a new Sub Account. By default, the Sub Account's signing key MUST be managed by the wallet. An application MAY optionally provide a set of signing keys (`keys`) for the Sub Account. A wallet SHOULD make the universal account an owner of the account, creating a hierarchical relationship between the newly created account and the universal account.

```typescript
type Parameters = {
  address: never;
  // Optional: keys of the account to be created
  keys?: { 
    publicKey: "0x...";
    type: "address" | "p256" | "webcrypto-p256" |  "webauthn-p256";
  }[];
  factory: never;
  factoryData: never;
}
```
```

Author: @jxon +1 to this improvement. This will make usage easier for developers.

status: Draft
type: Standards Track
category: ERC
created: 2025-02-18
---

## Abstract

This ERC introduces a new wallet RPC, `wallet_addSubAccount`, which allows an app to request a wallet to track a smart account that the wallet owns. It also allows apps to request the wallet to provision a new account, owned by the universal wallet with a signer provided by the caller. 

## Motivation

Embedded app accounts (onchain accounts specific to a single app) have led to a proliferation of user addresses, which can be difficult for users to keep track of. Many embedded app account users also have a universal wallet, which can be used across apps. With hierarchical ownership–where one smart account can own another–if the embedded app account is a smart account, it could be owned by the user’s universal wallet. This would allow users to be able to control an app account via their universal wallet. However, though hierarchical ownership is already possible today, there is no way for apps to tell universal wallets about embedded app accounts a user may have. The proposed RPC provides a path for this. 

## Specification

### Definitions

> Account - In this document, “account” means smart account. A smart contract that users transact from.
> Sub Account - An account that SHOULD have the main account as an owner of the sub-account. For instance, Account B is a sub-account of Account A if Account A is an owner, i.e. is a signer for Account B.
> 
> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

Reviewer: ```suggestion
The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

Account - In this document, “account” means smart account. A smart contract that users transact from.
Sub Account - An account that SHOULD have the main account as an owner of the sub-account. For instance, Account B is a sub-account of Account A if Account A is an owner, i.e. is a signer for Account B.
```

### JSON-RPC Methods

#### `wallet_addSubAccount`

> The `wallet_addSubAccount` RPC method allows applications to request that the connected account tracks the app account, creating a hierarchy between a universal account and app-embedded accounts. This RPC supports three different use cases.

Reviewer: You should definitely specify what "track" means. It's undefined.

Author: Thanks I'll define more clearly and update

Method names explored `wallet_linkAccount`,  `wallet_importAddress`, `wallet_addAddress`, or something else. Multiple RPCs were explored as well, separating create and tracking as two separate RPCs. To consolidate on a single RPC `wallet_addSubAccount` was selected. This method is more inclusive for EOA use cases. 

## Backwards Compatibility

This standard builds on existing JSON-RPC methods and complements [ERC-5792](./erc-5792.md) for future extensibility. Wallets can continue supporting legacy methods.

## Security Considerations 

As more capabilities are added, care should be taken to avoid unpredictable interactions. App specific accounts pose more risk for assets in those accounts. Having a universal account that maintains access to these accounts gives additional security to users and their funds.

### Privacy Considerations

Account data and any shared capabilities must be handled securely to avoid data leaks or man-in-the-middle attacks.

## Copyright

> Copyright and related rights waived via CC0.  

Reviewer: Please use the exact waiver from the template:

```suggestion
Copyright and related rights waived via [CC0](../LICENSE.md).
```
